Update DevTools to use getCacheForType API

bvaughn opened the pull request:
DevTools was built with a fork of an early idea for how Suspense cache might work. This idea is incompatible with newer APIs like `useTransition` which unfortunately prevented me from making certain UX improvements. This PR swaps out the primary usage of this cache (there are a few) in favor of the newer `unstable_getCacheForType` and `unstable_useCacheRefresh` APIs. We can go back and update the others in follow up PRs.

### Messaging changes

I've refactored the way the frontend loads component props/state/etc to hopefully make it better match the Suspense+cache model. Doing this gave up some of the small optimizations I'd added but hopefully the actual performance impact of that is minor and the overall ergonomic improvements of working with the cache API make this worth it.

The backend no longer remembers inspected paths. Instead, the frontend sends them every time and the backend sends a response with those paths. I've also added a new "force" parameter that the frontend can use to tell the backend to send a response even if the component hasn't rendered since the last time it asked. (This is used to get data for newly inspected paths.)

_Initial inspection..._
```
front |                                                      | back
      | -- "inspect" (id:1, paths:[], force:true) ---------> |
      | <------------------------ "inspected" (full-data) -- |
```
_1 second passes with no updates..._
```
      | -- "inspect" (id:1, paths:[], force:false) --------> |
      | <------------------------ "inspected" (no-change) -- |
```
_User clicks to expand a path, aka hydrate..._
```
      | -- "inspect" (id:1, paths:['foo'], force:true) ----> |
      | <------------------------ "inspected" (full-data) -- |
```
_1 second passes during which there is an update..._
```
      | -- "inspect" (id:1, paths:['foo'], force:false) ---> |
      | <----------------- "inspectedElement" (full-data) -- |
```

### Clear errors/warnings transition
Previously this meant there would be a delay after clicking the "clear" button. The UX after this change is much nicer:
![Suspense transition while clearing errors and warnings](https://user-images.githubusercontent.com/29597/103670753-ffee1f00-4f47-11eb-990d-2e741aaceecb.gif)

### Hydrating paths transition
I also added a transition to hydration (expanding "dehyrated" paths):
![Suspense transition while "hydrating" properties](https://user-images.githubusercontent.com/29597/104737246-02076900-5712-11eb-82c2-71afd3cf448f.gif)

### Better error boundaries
I also added a lower-level error boundary in case the new suspense operation ever failed. It provides a better "retry" mechanism (select a new element) so DevTools doesn't become entirely useful. Here I'm intentionally causing an error every time I select an element:

![Error boundary added around inspected element panel](https://user-images.githubusercontent.com/29597/104506720-c48cc980-55b3-11eb-91ab-bd2aa753ecdb.gif)

### Improved snapshot tests
I also migrated several of the existing snapshot tests to use inline snapshots and added a new serializer for dehydrated props. Inline snapshots are easier to verify and maintain and the new serializer means dehydrated props will be formatted in a way that makes sense rather than being empty (in external snapshots) or super verbose (default inline snapshot format).

Before (in `__snapshots__/inspectedElementContext-test.js.snap`)
![Screen Shot 2021-01-15 at 10 11 52 AM](https://user-images.githubusercontent.com/29597/104743826-21a28f80-571a-11eb-9f50-278037c5ab0c.png)

After (inline)
<img width="398" alt="Screen Shot 2021-01-15 at 10 09 17 AM" src="https://user-images.githubusercontent.com/29597/104743574-d8524000-5719-11eb-8e64-756b00bb547d.png">

gaearon commented on packages/react-devtools-shared/src/devtools/views/Components/InspectedElementContext.js:
```diff
@@ -145,14 +129,37 @@ function InspectedElementContextController({children}: Props) {
+        if (maybeInspectedElement !== undefined) {
+          return maybeInspectedElement;
+        } else {
+          const maybeThenable = inspectedElementThenables.get(element);
```
I think this should also live in your cache. E.g. you can structure the cache as `Map { key => record }` where `record` is `{ status, value }` and `value` is either the result (when status is fulfilled) or error (for rejected) or thenable (for pending). See `react-fs` for example.

bvaughn replied:
Yeah, agreed. I need to do some cleanup here. This was just a first pass to see how the transition works. :)

(Note this is a DRAFT RP still.)

gaearon commented on packages/react-devtools-shared/src/devtools/views/Components/InspectedElementContext.js:
```diff
@@ -277,16 +284,17 @@ function InspectedElementContextController({children}: Props) {
+            if (thenable != null) {
+              inspectedElementThenables.delete(element);
+
               batchedUpdates(() => {
```
This is a noop in CM?

gaearon commented on packages/react-devtools-shared/src/devtools/views/Components/InspectedElementContext.js:
```diff
@@ -198,7 +205,7 @@ function InspectedElementContextController({children}: Props) {
               fillInPath(inspectedElement, data.value, data.path, value);
 
-              resource.write(element, inspectedElement);
+              inspectedElementCache.set(element, inspectedElement);
```
Does this happen if the item already exists in the cache? If yes this looks like a mutation and I'm not sure that's valid.

bvaughn replied:
Seems to work ok based on my testing.

bvaughn replied:
But might still be a bad idea :)

gaearon replied:
I guess I don't know if it's any different from reading from a mutable store in general (if it's different, how?)

bvaughn replied:
Yeah, I need to think about this some more. I've got it listed as one of the TODOs on the PR. I just kind of swapped in place as much as possible with my initial commit. Good call to point it out though. 👍

bvaughn replied:
I guess the reason this "works" is that I'm duplicating the inspected element in local component state, and when these updates come in– I'm merging the update into both cache and state. (State is the thing I actually pass down through the context though.)

This was done originally to provide a mechanism to tell React that it needed to re-render after an inspected element update came from the backend. Happy to chat about alternate designs for this if you have ideas! :)

bvaughn commented:
Tests were failing b'c we never enabled cache for experimental test renderer builds (so the hook was missing).

bvaughn commented:
Cleaned this up a bit. My component organization here is still a little janky (e.g. the reason I have the memo-busting counter). I could refactor to move the context lower, within the suspense boundary, so that I could just cache on the object itself, but I noticed a lingering update that never completes (just re-renders at an interval) the first time I use `startTransition` to clear a warning.

Maybe I should do a more heavy refactor and try to remove this context entirely. _Something_ needs to poll for updates and refresh the cache but it doesn't necessarily need to be a context provider that sits above the Suspense boundary.

I may still be missing something about how the pieces for this kind of cache fit together. It's more context than our fs/db cache implementations, since there are a couple of types of invalidation. In general it seems like our `useCacheRefresh` API hasn't really been used for anything yet, outside of a couple of tests.

bvaughn commented:
I've refactored the way the frontend loads component props/state/etc to hopefully make it better match the Suspense+cache model. Doing this gave up some of the small optimizations I'd added but hopefully the actual performance impact of that is minor and the overall ergonomic improvements of working with the cache API make this worth it.

The backend no longer remembers inspected paths. Instead, the frontend sends them every time and the backend sends a response with those paths. I've also added a new "force" parameter that the frontend can use to tell the backend to send a response even if the component hasn't rendered since the last time it asked. (This is used to get data for newly inspected paths.)

Here's an example scenario:

_Initial inspection_
```
front |                                                      | back
      | -- "inspect" (id:1, paths:[], force:true) ---------> |
      | <------------------------ "inspected" (full-data) -- |
```
_1 second passes with no updates..._
```
      | -- "inspect" (id:1, paths:[], force:false) --------> |
      | <------------------------ "inspected" (no-change) -- |
```
_User clicks to expand a path, aka hydrate..._
```
      | -- "inspect" (id:1, paths:['foo'], force:true) ----> |
      | <------------------------ "inspected" (full-data) -- |
```
_1 second passes during which there is an update..._
```
      | -- "inspect" (id:1, paths:['foo'], force:false) ---> |
      | <----------------- "inspectedElement" (full-data) -- |
```

bvaughn commented:
All tests in `inspectedElement-test` pass now except for "_should poll for updates for the currently selected element_". That one's interesting because it seems like it _should_ be passing, except the transition+refresh that get scheduled never gets rendered. Will dig in more this afternoon.

bvaughn commented:
The only test that's failing now is this one:
```js
it('should poll for updates for the currently selected element', async done => {
  const Example = () => null;

  const container = document.createElement('div');
  await utils.actAsync(
    () => ReactDOM.render(<Example a={1} b="abc" />, container),
    false,
  );

  const id = ((store.getElementIDAtIndex(0): any): number);

  let inspectedElement = null;

  function Suspender({target}) {
    inspectedElement = useInspectedElement(id);
    return null;
  }

  let renderer;

  await utils.actAsync(() => {
    renderer = TestRenderer.create(
      <Contexts defaultSelectedElementID={id} defaultSelectedElementIndex={0}>
        <React.Suspense fallback={null}>
          <Suspender target={id} />
        </React.Suspense>
      </Contexts>,
    );
  }, false);
  expect(inspectedElement.props).toMatchInlineSnapshot(`
    Object {
      "a": 1,
      "b": "abc",
    }
  `);

  await utils.actAsync(
    () => ReactDOM.render(<Example a={2} b="def" />, container),
    false,
  );

  // Wait for our check-for-updates poll to get the new data.
  jest.runOnlyPendingTimers();
  await Promise.resolve();

  expect(inspectedElement.props).toMatchInlineSnapshot(`
    Object {
      "a": 2,
      "b": "def",
    }
  `);

  done();
});
```

The bit that waits for the check-for-updates poll works, in that– new data comes in and I refresh the cache, but that update never goes anywhere. React doesn't re-render `<Suspender>` afterward and so my test fails.

bvaughn commented:
Okay. That test is still not great and I'd love to get input on how to resolve it. Otherwise I think this PR is ready for another set of eyes!

acdlite reviewed (approved):
Neat! I only reviewed the part directly related to the cache API. I don't think I can comment on the rest of it because I don't know the DevTools codebase that well. Lemme know if you want a more thorough review.

eps1lon commented:
> That test is still not great and I'd love to get input on how to resolve it.

@bvaughn Does https://github.com/facebook/react/pull/20618 help?

bvaughn commented:
Thanks @acdlite! I wish I knew why [that one test](https://github.com/facebook/react/pull/20548#issuecomment-761205309) required such approach to pass, but I think that's something I can follow up with later.

@eps1lon Kind of! So the real issue was the `async -> done()` callback? (Obviously the date formatting issue but that's separate.) Maybe we can discuss the pros and cons of that in a follow up PR :)

bvaughn commented:
> > That test is still not great and I'd love to get input on how to resolve it.
> 
> @bvaughn Does https://github.com/facebook/react/pull/20618 help?

@eps1lon That comment was referring to [this](https://github.com/facebook/react/pull/20548#issuecomment-761205309) which is different.

Your findings with the inline snapshot are interesting though. So the real issue was the `async -> done()` callback? (Obviously the date formatting issue but that's separate.) Maybe we can discuss the pros and cons of that in a follow up PR :)
